File: submissions/0820-find-eventual-safe-states/solution.java
class Solution {
    public int dfs(int val,int[][] graph,int[] visited,int[] safe){
        visited[val] = 1;
        safe[val] = 1;
        int[] curr = graph[val];
        for(int j : curr){
            if(safe[j] == 1) return -1;
            else if(visited[j] == 1) continue;
            int value = dfs(j, graph,visited,safe);
            if(value == -1) return -1;
        }
        safe[val] = 0;
        return 0;
    }
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int[] visited = new int[graph.length];
        //int[] pathVisited = new int[graph.length];
        int[] safe = new int[graph.length];
        for(int i=0; i<graph.length; i++){
            if(visited[i] == 1) continue;
            dfs(i, graph,visited,safe);
        }
        List<Integer> ans = new ArrayList<>();
        for(int i=0; i<safe.length; i++){
            if(safe[i] == 0) ans.add(i);
        }
        return ans;
    }
}

File: submissions/0820-find-eventual-safe-states/solution.js
/**
 * @param {number[][]} graph
 * @return {number[]}
 */
var eventualSafeNodes = function(graph) {
    function dfs(val){
        if(pathVisited[val] === 1 || safe[val] === -1) return -1;
        if(safe[val] === 1) return 1;
        visited[val] = 1;
        pathVisited[val] = 1;
        let curr = graph[val];
        let count = 0;
        for(let j of curr){
            if(visited[j] === 1){
                count += safe[j];
                continue;
            }
            let value = dfs(j);
            if(value === -1) break;
            count += value;
        }
        if(count === curr.length) safe[val] = 1;
        else safe[val] = -1;
        pathVisited[val] = 0;
        return safe[val];
    }
    let n = graph.length;
        let visited = new Array(n);
        let pathVisited = new Array(n);
        let safe = new Array(n);
        for(let i=0; i<graph.length; i++){
            if(visited[i] === 1) continue;
            dfs(i);
        }
        let ans = [];
        for(let i=0; i<safe.length; i++){
            if(safe[i] === 1) ans = [...ans,i];
        }
        return ans;
};
